### PR TITLE
GRU/RNN/LSTM sequence ops, reference implementations, single layer tests

### PR DESCRIPTION
**Pull request by itikhono**
ngraph ops: GRU/RNN Sequence
references: GRU/RNN/LSTM Sequences
ngraph ie ops: GRU/RNN/LSTM SequenceIE
transformations: GRU/RNN/LSTM Sequence To SequenceIE, Bidirectional sequences to reverse-forward
unit tests, single layer tests

### CODE REVIEW

File: ngraph/core/src/pass/manager.cpp [hunk 1]
```diff
@@ -35,7 +35,7 @@ using namespace std;
 using namespace ngraph;
 
 pass::Manager::Manager()
-    : m_visualize(getenv_bool("NGRAPH_ENABLE_VISUALIZE_TRACING"))
+    : m_visualize(true)
```
Comment (lazarevevgeny) [inline ngraph/core/src/pass/manager.cpp:38]: Please, revert this change.

File: ngraph/core/include/ngraph/opsets/opset4_tbl.hpp [hunk 2]
```diff
@@ -155,10 +155,14 @@ NGRAPH_OP(TopK, ngraph::op::v3)
 NGRAPH_OP(Acosh, ngraph::op::v3)
 NGRAPH_OP(Asinh, ngraph::op::v3)
 NGRAPH_OP(Atanh, ngraph::op::v3)
+NGRAPH_OP(GRUSequence, ngraph::op::v4)
+NGRAPH_OP(LSTMSequence, ngraph::op::v0)
 NGRAPH_OP(CTCLoss, ngraph::op::v4)
 NGRAPH_OP(HSwish, ngraph::op::v4)
 NGRAPH_OP(NonMaxSuppression, ngraph::op::v4)
 NGRAPH_OP(Mish, ngraph::op::v4)
+NGRAPH_OP(CTCLoss, ngraph::op::v4)
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/opsets/opset4_tbl.hpp:164]: CTCLoss is mentioned twice. Please, fix this and make the operations in the alphabet order.

File: ngraph/core/include/ngraph/op/gru_sequence.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+#include "ngraph/op/op.hpp"
+#include "ngraph/op/util/rnn_cell_base.hpp"
+
+namespace ngraph
+{
+    namespace op
+    {
+        namespace v4
+        {
+            class NGRAPH_API GRUSequence : public Op, public util::RNNCellBase
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/op/gru_sequence.hpp:32]: May be we should change the inheritance scheme and make `util::RNNCellBase` to be inherited from `Op` class?

@ilyachur , @GlebKazantaev , what do you think about this? Will it break backward compatibility?

File: inference-engine/src/vpu/graph_transformer/src/frontend/unroll_loops.cpp [hunk 4]
```diff
@@ -34,12 +34,12 @@ void FrontEnd::unrollLoops(ie::ICNNNetwork& network) {
         return;
     }
 
-    if (env.config.enableTensorIteratorUnrolling) {
-        ie::NetPass::UnrollTI(network);
-    } else {
-        // Try to convert network to a RNN sequence due to performance reasons
-        ie::NetPass::CombineRNNSeq(network);
-    }
+//    if (env.config.enableTensorIteratorUnrolling) {
```
Comment (lazarevevgeny) [inline inference-engine/src/vpu/graph_transformer/src/frontend/unroll_loops.cpp:37]: May be remove these lines?

File: inference-engine/tests/functional/inference_engine/ngraph_reader/ti.cpp [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 #include <string>
 #include "ngraph_reader_tests.hpp"
-TEST_F(NGraphReaderTests, ReadTensorIteratorNetwork) {
+TEST_F(NGraphReaderTests, DISABLED_ReadTensorIteratorNetwork) {
```
Comment (lazarevevgeny) [inline inference-engine/tests/functional/inference_engine/ngraph_reader/ti.cpp:7]: Is it ok that the tests are disabled?

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp [hunk 6]
```diff
@@ -526,6 +529,132 @@ InferenceEngine::details::CNNLayerCreator::CNNLayerCreator(const std::shared_ptr
         return res;
     });
 
+    addSpecificCreator({"LSTMSequence"}, [](const std::shared_ptr<::ngraph::Node>& node,
```
Comment (ilyachur) [inline inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:532]: Why do you add such creators?

In general case we control the process of conversion. And plugins call all necessary transformations.
I am not sure that you can add tests for such creators without any hacks.

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp [hunk 7]
```diff
@@ -526,6 +529,132 @@ InferenceEngine::details::CNNLayerCreator::CNNLayerCreator(const std::shared_ptr
...
+                                                 const std::map<std::string, std::string> params) -> CNNLayerPtr {
```
Comment (ilyachur) [inline inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:533]: May I ask you to pass this map by reference?

And fix CreatorFor functor.

File: ngraph/core/include/ngraph/op/gru_sequence.hpp (hunk 3 above)
Comment (ilyachur) [reply ngraph/core/include/ngraph/op/gru_sequence.hpp:32]: Yes, I guess we should change the inheritance.

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp (hunk 6 above)
Comment (itikhono) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:532]: I think they are quite useful, if someone has deleted or forgot to add the necessary transformation, then this will help quickly understand the problem.
It is very easy to create a test - just don't call LSTMSequenceToLSTMSequenceIE transformation.

Comment (ilyachur) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:532]: In this case:

- First of all, could you add a test?
- The second one, maybe we can provide a special macros in order to generates such creators which throw exception. All these creators have the same code.

Comment (itikhono) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:532]: ok

File: inference-engine/src/transformations/include/ngraph_ops/gru_sequence_ie.hpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+#include "ngraph/opsets/opset4.hpp"
+#include "ngraph/op/op.hpp"
+
+namespace ngraph {
+namespace op {
+class TRANSFORMATIONS_API GRUSequenceIE : public Op, public ngraph::op::util::RNNCellBase {
+public:
+    GRUSequenceIE(const Output <Node> &X,
+                  const Output <Node> &H_t,
+                  // const Output <Node> &sequence_lengths,
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/ngraph_ops/gru_sequence_ie.hpp:22]: If its not needed lets remove it.

File: inference-engine/src/transformations/include/ngraph_ops/gru_sequence_ie.hpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+                  const Output <Node> &B,
+                  size_t hidden_size,
+                  op::RecurrentSequenceDirection direction,
+                  const std::vector<std::string> &activations,
+                  const std::vector<float> &activations_alpha,
+                  const std::vector<float> &activations_beta,
+                  float clip,
+                  bool linear_before_reset);
+
+    static constexpr NodeTypeInfo type_info{"GRUSequenceIE", 4};
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/ngraph_ops/gru_sequence_ie.hpp:33]: Please use new NGRAPH_RTTI_DECLARATION/DEFINITION

File: inference-engine/src/transformations/include/ngraph_ops/lstm_sequence_ie.hpp [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+#include <transformations_visibility.hpp>
+
+#include "ngraph/opsets/opset4.hpp"
+#include "ngraph/op/op.hpp"
+
+namespace ngraph {
+namespace op {
+class TRANSFORMATIONS_API LSTMSequenceIE : public Op, public ngraph::op::util::RNNCellBase {
+public:
+    static constexpr NodeTypeInfo type_info{"LSTMSequenceIE", 1};
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/ngraph_ops/lstm_sequence_ie.hpp:20]: NGRAPH_RTTI_DECLARATION/DEFINITION

File: inference-engine/src/transformations/include/ngraph_ops/rnn_sequence_ie.hpp [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+#include "ngraph/opsets/opset4.hpp"
+#include "ngraph/op/op.hpp"
+
+namespace ngraph {
+namespace op {
+class TRANSFORMATIONS_API RNNSequenceIE : public Op, public ngraph::op::util::RNNCellBase {
+public:
+    RNNSequenceIE(const Output <Node> &X,
+                  const Output <Node> &H_t,
+                  //const Output <Node> &sequence_lengths,
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/ngraph_ops/rnn_sequence_ie.hpp:22]: lets remove if it is not needed

File: inference-engine/src/transformations/include/ngraph_ops/rnn_sequence_ie.hpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+                  const Output <Node> &WR,
+                  const Output <Node> &B,
+                  size_t hidden_size,
+                  op::RecurrentSequenceDirection direction,
+                  const std::vector<std::string> &activations,
+                  const std::vector<float> &activations_alpha,
+                  const std::vector<float> &activations_beta,
+                  float clip);
+
+    static constexpr NodeTypeInfo type_info{"RNNSequenceIE", 4};
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/ngraph_ops/rnn_sequence_ie.hpp:32]: NGRAPH_RTTI_DECLARATION/DEINITION

File: inference-engine/src/transformations/include/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.hpp [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+class TRANSFORMATIONS_API ConvertLSTMSequenceMatcher;
+class TRANSFORMATIONS_API ConvertGRUSequenceMatcher;
+class TRANSFORMATIONS_API ConvertRNNSequenceMatcher;
+
+}  // namespace pass
+}  // namespace ngraph
+
+/**
+ * @ingroup ie_transformation_common_api
+ * @brief Converts opset4::(GRU/RNN/LSTM)Sequence to legacy (GRU/RNN/LSTM)SequenceIE.
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.hpp:27]: Please add such comment for every transformation class.

File: inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/convert_tensor_iterator_to_sequence.hpp [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+    class TRANSFORMATIONS_API ConvertTensorIteratorToLSTMSequence;
+    class TRANSFORMATIONS_API ConvertTensorIteratorToRNNSequence;
+    class TRANSFORMATIONS_API ConvertTensorIteratorToGRUSequence;
+
+}  // namespace pass
+}  // namespace ngraph
+
+/**
+ * @ingroup ie_transformation_common_api
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/convert_tensor_iterator_to_sequence.hpp:28]: Add description for every transformation class

File: inference-engine/src/transformations/src/ngraph_ops/gru_sequence_ie.cpp [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+          m_direction(direction),
+          m_linear_before_reset(linear_before_reset) {
+    constructor_validate_and_infer_types();
+}
+
+void op::GRUSequenceIE::validate_and_infer_types() {
+    element::Type arg_type = get_input_element_type(0);
+    PartialShape output_shape_0{PartialShape::dynamic(3)};
+    PartialShape output_shape_1{PartialShape::dynamic(2)};
+    if (get_input_partial_shape(0).is_static()) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/ngraph_ops/gru_sequence_ie.cpp:38]: Lets check that rank is static and greater that 2 and 0/1 dims are static and then get them for batch_size and seq_length

File: inference-engine/src/transformations/src/ngraph_ops/lstm_sequence_ie.cpp [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+          RNNCellBase(hidden_size, clip, activations, activations_alpha, activations_beta),
+          m_direction(direction) {
+    constructor_validate_and_infer_types();
+}
+
+void op::LSTMSequenceIE::validate_and_infer_types() {
+    element::Type arg_type = get_input_element_type(0);
+    PartialShape output_shape_0{PartialShape::dynamic(3)};
+    PartialShape output_shape_1{PartialShape::dynamic(2)};
+    if (get_input_partial_shape(0).is_static()) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/ngraph_ops/lstm_sequence_ie.cpp:37]: the same comment as above

File: inference-engine/src/transformations/src/ngraph_ops/rnn_sequence_ie.cpp [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+          m_direction(direction) {
+    constructor_validate_and_infer_types();
+}
+
+void op::RNNSequenceIE::validate_and_infer_types() {
+    element::Type arg_type = get_input_element_type(0);
+    PartialShape output_shape_0{PartialShape::dynamic(4)};
+    PartialShape output_shape_1{PartialShape::dynamic(3)};
+    if (get_input_partial_shape(0).is_static()) {
+        size_t batch_size = get_input_partial_shape(0).get_shape()[0];
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/ngraph_ops/rnn_sequence_ie.cpp:37]: the same comment as above

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+
+        auto W = std::dynamic_pointer_cast<ngraph::opset4::Constant>(
+                lstm_sequence->input_value(4).get_node_shared_ptr());
+        if (!W) {
+            return false;
+        }
+
+        auto R = std::dynamic_pointer_cast<ngraph::opset4::Constant>(
+                lstm_sequence->input_value(5).get_node_shared_ptr());
+        if (!R) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp:34]: please move this checks to pattern:
```suggestion
// JUST EXAMPLE, PLEASE USE NORMAL NAMES FOR VARIABLES
auto input1 = pattern::any_input();
...
auto W = ngraph::pattern::wrap_type<ngraph::opset4::Constant>();
auto lstm_sequence_ngraph = ngraph::pattern::wrap_type<ngraph::opset4::LSTMSequence>(input1, input2, W, R);
```

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+
+        for (int out_id = 0; out_id < lstm_sequence->outputs().size(); ++out_id) {
+            for (const auto& target_input : lstm_sequence->output(out_id).get_target_inputs()) {
+                target_input.replace_source_output(unsqueezes[out_id]);
+            }
+        }
+        lstm_sequence_ie->set_friendly_name(lstm_sequence->get_friendly_name());
+        ngraph::copy_runtime_info(lstm_sequence, {concat, lstm_sequence_ie, in_1, in_2, in_3, in_4, unsqueeze_1,
+                                                  unsqueeze_2, unsqueeze_3});
+        ngraph::replace_node(m.get_match_root(), lstm_sequence_ie);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp:74]:
```suggestion
        ngraph::replace_node(lstm_sequence, lstm_sequence_ie);
```

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+                lstm_sequence->get_clip_threshold());
+
+        auto unsqueeze_axis = ngraph::opset4::Constant::create(ngraph::element::i64, ngraph::Shape{1}, {1});
+        auto unsqueeze_1 = std::make_shared<ngraph::opset4::Unsqueeze>(lstm_sequence_ie->output(0), unsqueeze_axis);
+        auto unsqueeze_2 = std::make_shared<ngraph::opset4::Unsqueeze>(lstm_sequence_ie->output(1), unsqueeze_axis);
+        auto unsqueeze_3 = std::make_shared<ngraph::opset4::Unsqueeze>(lstm_sequence_ie->output(2), unsqueeze_axis);
+        NodeVector unsqueezes = {unsqueeze_1, unsqueeze_2, unsqueeze_3};
+
+        for (int out_id = 0; out_id < lstm_sequence->outputs().size(); ++out_id) {
+            for (const auto& target_input : lstm_sequence->output(out_id).get_target_inputs()) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp:67]: replace this cycle with 
```suggestion
lstm_sequence->output(out_id).replace(unsqueezes[out_id]);
```

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+        auto axis_1 = ngraph::opset4::Constant::create(ngraph::element::i64, ngraph::Shape{1}, {1});
+        auto in_1 = std::make_shared<ngraph::opset4::Squeeze>(rnn_sequence->input(1).get_source_output(), axis_1);
+        auto concat = std::make_shared<ngraph::opset4::Concat>(ngraph::NodeVector({W, R}), 2);
+        auto axis_2 = ngraph::opset4::Constant::create(ngraph::element::i64, ngraph::Shape{1}, {0});
+        auto in_3 = std::make_shared<ngraph::opset4::Squeeze>(concat->output(0), axis_2);
+        auto in_4 = std::make_shared<ngraph::opset4::Squeeze>(rnn_sequence->input(5).get_source_output(), axis_2);
+        auto rnn_sequence_ie = std::make_shared<ngraph::op::RNNSequenceIE>(
+                rnn_sequence->input(0).get_source_output(),  // X
+                in_1,  // initial_hidden_state
+//                rnn_sequence->input(2).get_source_output(),  // sequence_lengths
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp:179]: to remove?

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+                rnn_sequence->get_clip());
+
+        auto unsqueeze_axis = ngraph::opset4::Constant::create(ngraph::element::i64, ngraph::Shape{1}, {1});
+        auto unsqueeze_1 = std::make_shared<ngraph::opset4::Unsqueeze>(rnn_sequence_ie->output(0), unsqueeze_axis);
+        auto unsqueeze_2 = std::make_shared<ngraph::opset4::Unsqueeze>(rnn_sequence_ie->output(1), unsqueeze_axis);
+        NodeVector unsqueezes = {unsqueeze_1, unsqueeze_2};
+
+        for (int out_id = 0; out_id < rnn_sequence->outputs().size(); ++out_id) {
+            for (const auto& target_input : rnn_sequence->output(out_id).get_target_inputs()) {
+                target_input.replace_source_output(unsqueezes[out_id]);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp:196]: the same comment as above

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+            }
+        }
+        rnn_sequence_ie->set_friendly_name(rnn_sequence->get_friendly_name());
+        ngraph::copy_runtime_info(rnn_sequence, {concat, rnn_sequence_ie, in_1, in_3, in_4, unsqueeze_1,
+                                                 unsqueeze_2});
+        ngraph::replace_node(m.get_match_root(), rnn_sequence_ie);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp:202]:
```suggestion
        ngraph::replace_node(rnn_sequence_ie, rnn_sequence_ie);
```

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp [hunk 24]
```diff
@@ -0,0 +1,208 @@
+// Copyright (C) 2020 Intel Corporation
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp:1]: Please move this transformations files to convert_opset1_to_legacy

File: inference-engine/src/vpu/graph_transformer/src/frontend/frontend.cpp [hunk 25]
```diff
@@ -395,14 +398,27 @@ ModelPtr FrontEnd::runCommonPasses(ie::ICNNNetwork& network, const UnsupportedLa
             ngraph::op::GenericIE::DisableReshape noReshape(nGraphFunc);
 
             ngraph::pass::Manager manager;
+            const auto& env = CompileEnv::get();
+            if (!env.config.forcePureTensorIterator) {
```
Comment (GlebKazantaev) [inline inference-engine/src/vpu/graph_transformer/src/frontend/frontend.cpp:402]: let's do not include this passes into CommonOptimization until we have not mechanism for transformations disabling.

File: ngraph/core/include/ngraph/op/gru_sequence.hpp [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+            {
+            public:
+                static constexpr NodeTypeInfo type_info{"GRUSequence", 4};
```
Comment (GlebKazantaev) [inline ngraph/core/include/ngraph/op/gru_sequence.hpp:35]: NGRAPH_RTTI_DEFINITION

File: ngraph/core/include/ngraph/op/rnn_sequence.hpp [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+namespace ngraph
+{
+    namespace op
+    {
+        namespace v4
+        {
+            class NGRAPH_API RNNSequence : public Op, public util::RNNCellBase
+            {
+            public:
+                static constexpr NodeTypeInfo type_info{"RNNSequence", 4};
```
Comment (GlebKazantaev) [inline ngraph/core/include/ngraph/op/rnn_sequence.hpp:35]: NGRAPH_RTTI_DEFINITION

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Comment by GlebKazantaev** [comment]
Will continue review tommorow.

File: ngraph/core/include/ngraph/op/gru_sequence.hpp (hunk 3 above)
Comment (itikhono) [reply ngraph/core/include/ngraph/op/gru_sequence.hpp:32]: inputs are not passed to `util::RNNCellBase`, classes inherited from `util::RNNCellBase` can have different inputs, the class is only used to store common attributes.

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp [hunk 28]
```diff
@@ -102,6 +104,10 @@ InferenceEngine::ICNNNetwork::Ptr clDNNEngine::CloneAndTransformNetwork(const In
         // Note: instead of running all Conversion Transformations you can make up your own transformation pipeline
         ngraph::pass::Manager manager;
         manager.register_pass<ngraph::pass::CommonOptimizations>();
+        manager.register_pass<ngraph::pass::ConvertTensorIteratorToLSTMSequence>();
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/cldnn_engine/cldnn_engine.cpp:108]: We should use env options for some plugins to enable/disable these transformations, e.g vpu plugin

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_opset4_to_opset3.cpp [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+
+#include <ngraph/pass/manager.hpp>
+#include <ngraph/pass/constant_folding.hpp>
+
+bool ngraph::pass::ConvertOpSet4ToOpSet3::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    OV_ITT_SCOPED_TASK(itt::domains::IETransform, "ngraph::pass::ConvertOpSet4ToOpSet3");
+
+    ngraph::pass::Manager manager;
+
+    manager.register_pass<ngraph::pass::ConvertGRUSequenceMatcher>();
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_opset4_to_opset3.cpp:20]: is it ok that GRU/RNN Seq is part of opset4, not opset1?

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp (hunk 7 above)
Comment (itikhono) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:533]: done in a separate PR https://github.com/openvinotoolkit/openvino/pull/1928

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp (hunk 6 above)
Comment (itikhono) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:532]: done in a separate PR https://github.com/openvinotoolkit/openvino/pull/1928

File: ngraph/core/include/ngraph/op/gru_sequence.hpp (hunk 3 above)
Comment (itikhono) [reply ngraph/core/include/ngraph/op/gru_sequence.hpp:32]: done in the separate PR: https://github.com/openvinotoolkit/openvino/pull/2000

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_opset4_to_opset3.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_opset4_to_opset3.cpp]: moved to Opset1ToLegacy

File: inference-engine/tests/functional/inference_engine/ngraph_reader/ti.cpp (hunk 5 above)
Comment (itikhono) [reply inference-engine/tests/functional/inference_engine/ngraph_reader/ti.cpp]: these changes were moved to a separate branch

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp (hunk 28 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/cldnn_engine/cldnn_engine.cpp]: these changes were moved to a separate branch

File: inference-engine/src/transformations/include/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.hpp (hunk 13 above)
Comment (itikhono) [reply inference-engine/src/transformations/include/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.hpp]: done

File: inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/convert_tensor_iterator_to_sequence.hpp (hunk 14 above)
Comment (itikhono) [reply inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/convert_tensor_iterator_to_sequence.hpp]: these changes were moved to a separate branch

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp (hunk 19 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp]: done

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp (hunk 20 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp]: replaced on single replace_node() call

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp (hunk 23 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp]: done

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp (hunk 22 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp]: replaced with single replace_node() call

File: inference-engine/src/legacy_api/src/cnn_network_impl.cpp [hunk 30]
```diff
@@ -97,13 +100,20 @@ CNNNetworkImpl::CNNNetworkImpl(const ICNNNetwork & ngraphImpl) {
...
     manager.register_pass<::ngraph::pass::ConvertOpSet2ToOpSet1>();
     manager.register_pass<::ngraph::pass::ConvertOpSet1ToLegacy>();
     manager.run_passes(graph);
 
     ::ngraph::pass::Manager ti_manager;
-    ti_manager.register_pass<::ngraph::pass::ApplyTransformationsToTIBody>(manager);
+    // Apply all transformations to TensorIterator body
+    ti_manager.register_pass<ngraph::pass::ApplyTransformationsToTIBody>(manager);
+    // Unroll will be called after all conversions
+    ti_manager.register_pass<ngraph::pass::UnrollTensorIterator>();
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/legacy_api/src/cnn_network_impl.cpp]: these changes were moved to a separate branch, will be fixed

File: inference-engine/src/vpu/graph_transformer/src/frontend/frontend.cpp (hunk 25 above)
Comment (itikhono) [reply inference-engine/src/vpu/graph_transformer/src/frontend/frontend.cpp]: these changes were moved to a separate branch

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp (hunk 24 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp]: done

File: inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp (hunk 21 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/convert_opset4_to_opset3/convert_sequences_to_sequences_ie.cpp]: removed

File: inference-engine/src/transformations/include/ngraph_ops/rnn_sequence_ie.hpp (hunk 12 above)
Comment (itikhono) [reply inference-engine/src/transformations/include/ngraph_ops/rnn_sequence_ie.hpp]: done

**Review by iefode — COMMENTED** [review]
Test part LGTM, but similar comment as in https://github.com/openvinotoolkit/openvino/pull/2000

**Comment by GlebKazantaev** [comment]
Reviewed only ngraph part.

**Review by GlebKazantaev — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Comment by itikhono** [comment]
@jane-intel could you check the processing dynamic shapes in validate_and_infer functions of GRU/RNN/LSTM Sequences/Cells ops?

**Review by iefode — APPROVED** [review]

**Comment by lazarevevgeny** [comment]
@itikhono , please, resolve merge conflicts

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]
Update the operation specification

**Comment by itikhono** [comment]
> Update the operation specification

done, could you please verify?

**Review by lazarevevgeny — APPROVED** [review]
Reviewed the specification part.

**Review by mikhail-treskin — APPROVED** [review]

File: ngraph/core/src/op/rnn_sequence.cpp [hunk 31]
```diff
@@ -0,0 +1,175 @@
...
+    , m_direction(direction)
+{
+    constructor_validate_and_infer_types();
+}
+
+void op::v5::RNNSequence::validate_and_infer_types()
+{
+    for (const auto& input : inputs())
+    {
+        if (input.get_partial_shape().rank().is_dynamic())
```
Comment (jane-intel) [inline ngraph/core/src/op/rnn_sequence.cpp:62]: what if the rank is known but shape in partial? further `get_shape()` will fail

Comment (itikhono) [reply ngraph/core/src/op/rnn_sequence.cpp:62]: fixed

**Review by jane-intel — APPROVED** [review]
